### PR TITLE
[scudo] Add support for LoongArch hardware CRC32 checksumming

### PR DESCRIPTION
**Pull request by xen0n**
One has to probe for platform capability prior to use with HWCAP, according to LoongArch documentation.

### CODE REVIEW

**Comment by xen0n** [comment]
cc @SixWeining @heiher @xry111

File: compiler-rt/lib/scudo/standalone/checksum.cpp
Comment (xry111) [inline compiler-rt/lib/scudo/standalone/checksum.cpp:22]: Hmm, why not just __loongarch__?  Even if some LA32 implementations do not support CRC32 we can detect it with HWCAP anyway.

Comment (xen0n) [reply compiler-rt/lib/scudo/standalone/checksum.cpp:22]: I was following the manual word-for-word. But I think your suggestion is reasonable and I'll change it later.

Comment (xry111) [reply compiler-rt/lib/scudo/standalone/checksum.cpp:22]: The manual basically says "LA32 SHALL implement the instructions in table 2-1", but there is no clause saying "LA32 SHALL NOT implement other instructions".

Comment (xen0n) [reply compiler-rt/lib/scudo/standalone/checksum.cpp]: Done.

**Review by ChiaHungDuan — APPROVED** [review]
Thanks for adding this

File: compiler-rt/lib/scudo/standalone/checksum.cpp
Comment (SixWeining) [inline compiler-rt/lib/scudo/standalone/checksum.cpp:87]: I got this error:
```
compiler-rt/lib/scudo/standalone/checksum.cpp:87:35: error: use of undeclared identifier 'HWCAP_LOONGARCH_CRC32'
```

Maybe a `.h` file should be included?

**Review by SixWeining — COMMENTED** [review]
Thanks for the patch.

File: compiler-rt/lib/scudo/standalone/checksum.cpp
Comment (SixWeining) [reply compiler-rt/lib/scudo/standalone/checksum.cpp:87]: After `<asm/hwcap.h>` is included, I got this error:
```
compiler-rt/lib/scudo/standalone/crc32_hw.cpp:23:49: error: implicit conversion changes signedness: 'u32' (aka 'unsigned int') to 'int' [-Werror,-Wsign-conversion]
   23 |   return static_cast<u32>(CRC32_INTRINSIC(Data, Crc));
      |                           ~~~~~~~~~~~~~~~       ^~~
compiler-rt/lib/scudo/standalone/crc32_hw.cpp:23:43: error: implicit conversion changes signedness: 'uptr' (aka 'unsigned long') to 'long' [-Werror,-Wsign-conversion]
   23 |   return static_cast<u32>(CRC32_INTRINSIC(Data, Crc));
      |                           ~~~~~~~~~~~~~~~ ^~~~
2 errors generated.
```

Note that I'm using clang-17 to build.

Comment (xry111) [reply compiler-rt/lib/scudo/standalone/checksum.cpp:87]: > After `<asm/hwcap.h>` is included, I got this error:
> 
> ```
> compiler-rt/lib/scudo/standalone/crc32_hw.cpp:23:49: error: implicit conversion changes signedness: 'u32' (aka 'unsigned int') to 'int' [-Werror,-Wsign-conversion]
>    23 |   return static_cast<u32>(CRC32_INTRINSIC(Data, Crc));
>       |                           ~~~~~~~~~~~~~~~       ^~~
> compiler-rt/lib/scudo/standalone/crc32_hw.cpp:23:43: error: implicit conversion changes signedness: 'uptr' (aka 'unsigned long') to 'long' [-Werror,-Wsign-conversion]
>    23 |   return static_cast<u32>(CRC32_INTRINSIC(Data, Crc));
>       |                           ~~~~~~~~~~~~~~~ ^~~~
> 2 errors generated.
> ```
> 
> Note that I'm using clang-17 to build.

Yes, this is nasty.  `__crc_w_d_w` etc. has signed argument and return value types.  I had to add a lot of casts in my xz CRC32 optimization to silence the warning.

Comment (xen0n) [reply compiler-rt/lib/scudo/standalone/checksum.cpp:87]: > I got this error:
> 
> ```
> compiler-rt/lib/scudo/standalone/checksum.cpp:87:35: error: use of undeclared identifier 'HWCAP_LOONGARCH_CRC32'
> ```
> 
> Maybe a `.h` file should be included?

Hmm I was wrong in assuming the HWCAP constants are implicitly pulled in by `<sys/auxv.h>`. This is the case only [from glibc 2.38](https://github.com/bminor/glibc/commit/7f079fdc16e88ebb8020e17b2fd900e8924da29a). I'll supply the definition with an `#ifdef` check.

**Review by SixWeining — APPROVED** [review]

**Comment by SixWeining** [comment]
When I click the merge button, I get this prompt "This commit will be authored by 1175567+xen0n@users.noreply.github.com". Do you mind to change your email settings before this getting landed?

**Comment by SixWeining** [comment]
FYI: https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-personal-account-on-github/managing-email-preferences/setting-your-commit-email-address

**Pull request merged**
